solr/core/src/java/org/apache/solr/cli/CLIUtils.java

```diff
@@ -0,0 +1,357 @@
+   *
+   * @param solrUrl the URL to extract the port from
+   * @return The port that was found.
+   * @throws NullPointerException If solrUrl is null
+   * @throws URISyntaxException If the given string violates RFC 2396, as augmented by the above
+   *     deviations
+   */
+  public static int portFromUrl(String solrUrl) throws URISyntaxException {
+    URI uri = new URI(solrUrl);
+    int port = uri.getPort();
+    if (port == -1) {
+      return uri.getScheme().equals("https") ? 443 : 80;
+    } else {
+      return port;
+    }
+  }
+
+  public static boolean safeCheckCollectionExists(
+      String solrUrl, String collection, String credentials) {
+    boolean exists = false;
+    try (var solrClient = getSolrClient(solrUrl, credentials)) {
+      NamedList<Object> existsCheckResult = solrClient.request(new CollectionAdminRequest.List());
+      @SuppressWarnings("unchecked")
+      List<String> collections = (List<String>) existsCheckResult.get("collections");
+      exists = collections != null && collections.contains(collection);
+    } catch (Exception exc) {
+      // just ignore it since we're only interested in a positive result here
+    }
+    return exists;
+  }
+
+  @SuppressWarnings("unchecked")
+  public static boolean safeCheckCoreExists(String solrUrl, String coreName, String credentials) {
+    boolean exists = false;
+    try (var solrClient = getSolrClient(solrUrl, credentials)) {
+      boolean wait = false;
+      final long startWaitAt = System.nanoTime();
+      do {
+        if (wait) {
+          final int clamPeriodForStatusPollMs = 1000;
+          Thread.sleep(clamPeriodForStatusPollMs);
+        }
+        NamedList<Object> existsCheckResult =
+            CoreAdminRequest.getStatus(coreName, solrClient).getResponse();
+        NamedList<Object> status = (NamedList<Object>) existsCheckResult.get("status");
+        NamedList<Object> coreStatus = (NamedList<Object>) status.get(coreName);
+        Map<String, Object> failureStatus =
+            (Map<String, Object>) existsCheckResult.get("initFailures");
+        String errorMsg = (String) failureStatus.get(coreName);
+        final boolean hasName = coreStatus != null && coreStatus.asMap().containsKey(NAME);
+        exists = hasName || errorMsg != null;
+        wait = hasName && errorMsg == null && "true".equals(coreStatus.get("isLoading"));
+      } while (wait && System.nanoTime() - startWaitAt < MAX_WAIT_FOR_CORE_LOAD_NANOS);
+    } catch (Exception exc) {
+      // just ignore it since we're only interested in a positive result here
+    }
+    return exists;
+  }
+
+  public static boolean isCloudMode(SolrClient solrClient) throws SolrServerException, IOException {
+    NamedList<Object> systemInfo =
+        solrClient.request(new GenericSolrRequest(SolrRequest.METHOD.GET, SYSTEM_INFO_PATH));
+    return "solrcloud".equals(systemInfo.get("mode"));
+  }
+
+  public static Path getConfigSetsDir(Path solrInstallDir) {
```

Comment:
I believe it was previously in the CLI scripts, and there is one more reference to this directory in RunExampleTool that uses the server dir instead of the install dir:

```java
786 File configsetsDir = new File(serverDir, "solr/configsets");
```

If we move at some point all path computation to the CLI via options and defaults, we could probably improve and merge these parts. See for reference [SolrContextOptions.kt](https://github.com/malliaridis/solr/blob/7b295068f1c679aef54863fd35b679db94301561/solr/cli/src/main/kotlin/org/apache/solr/cli/options/SolrContextOptions.kt).